python/cudf_polars/cudf_polars/experimental/sort.py

```diff
@@ -4,19 +4,391 @@
 from cudf_polars.experimental.repartition import Repartition
-from cudf_polars.experimental.utils import _lower_ir_fallback
+from cudf_polars.experimental.shuffle import _simple_shuffle_graph
+from cudf_polars.experimental.utils import _concat, _lower_ir_fallback
 
 if TYPE_CHECKING:
-    from collections.abc import MutableMapping
+    from collections.abc import MutableMapping, Sequence
 
-    from cudf_polars.dsl.ir import IR
+    from cudf_polars.containers import DataType
+    from cudf_polars.dsl.expr import NamedExpr
     from cudf_polars.experimental.dispatch import LowerIRTransformer
+    from cudf_polars.typing import Schema
+    from cudf_polars.utils.config import ConfigOptions
+
+
+def find_sort_splits(
+    tbl: plc.Table,
+    sort_boundaries: plc.Table,
+    my_part_id: int,
+    column_order: Sequence[plc.types.Order],
+    null_order: Sequence[plc.types.NullOrder],
+) -> list[int]:
+    """
+    Find local sort splits given all (global) split candidates.
+
+    The reason for much of the complexity is to get the result sizes as
+    precise as possible even when e.g. all values are equal.
+    In other words, this goes through extra effort to split the data at the
+    precise boundaries (which includes part_id and local_row_number).
+
+    Parameters
+    ----------
+    tbl
+        Locally sorted table only containing sort columns.
+    sort_boundaries
+        Sorted table containing the global sort/split boundaries.  Compared to `tbl`
+        must contain additional partition_id and local_row_number columns.
+    my_part_id
+        The partition id of the local node (as the `split_candidates` column).
+    column_order
+        The order in which tbl is sorted.
+    null_order
+        The null order in which tbl is sorted.
+
+    Returns
+    -------
+    The split points for the local partition.
+    """
+    column_order = list(column_order)
+    null_order = list(null_order)
+
+    # We now need to find the local split points.  To do this, first split out
+    # the partition id and the local row number of the final split values
+    *sort_boundaries, split_part_id, split_local_row = sort_boundaries.columns()
+    sort_boundaries = plc.Table(sort_boundaries)
+    # Now we find the first and last row in the local table corresponding to the split value
+    # (first and last, because there may be multiple rows with the same split value)
+    split_first_col = plc.search.lower_bound(
+        tbl, sort_boundaries, column_order, null_order
+    )
+    split_last_col = plc.search.upper_bound(
+        tbl, sort_boundaries, column_order, null_order
+    )
+    # And convert to arrow/CPU for final processing
+    split_first_col = plc.interop.to_arrow(split_first_col).to_pylist()
+    split_last_col = plc.interop.to_arrow(split_last_col).to_pylist()
+    split_part_id = plc.interop.to_arrow(split_part_id).to_pylist()
+    split_local_row = plc.interop.to_arrow(split_local_row).to_pylist()
```

Comment:
Pointed out by @mroeschke offline, we should be able to do something like. So we do *not* need #19446 
```python
    split_first_col = pl.Series(split_first_col).to_list()
    split_last_col = pl.Series(split_last_col).to_list()
    split_part_id = pl.Series(split_part_id).to_list()
    split_local_row = pl.Series(split_local_row).to_list()
```